src/java/org/apache/cassandra/config/Config.java

```diff
@@ -280,6 +280,9 @@
     public volatile int tombstone_warn_threshold = 1000;
     public volatile int tombstone_failure_threshold = 100000;
 
+    public volatile int cached_replica_rows_warn_threshold = 1024;
+    public volatile int cached_replica_rows_fail_threshold = 16384;
```

Comment:
My intuition from a few years of index query testing is that we shouldn't go much higher on the warn threshold here. Both of these are subject to change though, and we should do some basic stress testing before we commit.

One potentially interesting interaction here is that it might be possible for the failure threshold to be set below the query limit. The default value is high enough here that it probably won't affect very many people at all, but it's worth mentioning.

Reply:
I don't know if using powers of two gives us any advantage here, and I'm a bit afraid it could make users thing that this is somehow measuring memory sizes. Given that this seems more related to the query limit, perhaps we should use something like 1000, WDYT?

Also, that is the number of rows in all replicas, so we are reading from two on-sync replicas the threshold will start at 512 results, won't it? Perhaps we should count the number of merged materialized/cached rows, ignoring the replicas, so the thresholds can be more easily related to the query limit.

Reply:
> perhaps we should use something like 1000, WDYT?

Yup. No particular attachment to powers of two.

> so if we are reading from two on-sync replicas the threshold will start at 512 results, allowing a max limit/page size of 512 rows, won't it

Correct.

> Perhaps we should count the number of merged materialized/cached rows, ignoring the replicas, so the thresholds can be more easily related to the query limit.

I agree that it would be a bit easier to relate that to the limit, but my worry is that it would make the same threshold values behave differently with different read consistency levels. That also brings up another question, which is whether, even in this scheme, we should count the merged row as well.

If we ultimately don't need this to be directly relatable to the limit, and I propose we don't (see `tombstone_failure_threshold`), it may also make sense to go down to the level of cells. It doesn't actually look much more expensive to get the number of cells from `Row#columnCount()`, but it adds some complexity (and doesn't look like it includes the clustering).

Reply:
>I agree that it would be a bit easier to relate that to the limit, but my worry is that it would make the same threshold values behave differently with different read consistency levels.

Makes sense. The other way around, I understand that our intention is to provide a default threshold value that keeps heap usage reasonably low while allowing a reasonably high page size for the most common number of replicas. I think that the number of replicas is two, since for one replica RFP is not used. In that case, a default threshold value of 1000 means that the max allowed page size without warnings will be 500. Perhaps we should use a threshold value of 2000 to use a more round number (1000) in the page size, or it's too high? I don't really now what can be considered the maximum reasonable page size. For example the Spark driver uses 1000 as default, but with `LOCAL_ONE` consistency so it's not affected by this.

>That also brings up another question, which is whether, even in this scheme, we should count the merged row as well. (I'd say we don't have to, given the merged row is already being kept on heap outside RFP.)

Agree, the merged rows are not cached but consumed from an iterator as usual, so I wouldn't count them.

>If we ultimately don't need this to be directly relatable to the limit, and I propose we don't (see tombstone_failure_threshold), it may also make sense to go down to the level of cells. It doesn't actually look much more expensive to get the number of cells from Row#columnCount(), but it adds some complexity (and doesn't look like it includes the clustering).

The size of the cells is so dependant on the data they contain as the rows, so I'm not sure that would give us more control about the amount of heap usage. In both cases, and while we don't count actual payload size, users are going to need to think about what data are they storing/querying to get an estimate of the memory pressure. Calculating it from rows or cells doesn't seem to make a big difference, although I think I'd prefer rows.

Reply:
If this is to be replica and page size dependent, perhaps it should be defined as a ratio of some multiple of these?

Reply:
The ratio sounds like a good idea for usability, since it allows to define how much overload are we willing to accept in a per-query basis. The downside is that it limits the capacity of administrators (who edit the yaml) to limit the damage that users can do if they decide to use a very high page size, I think.

Reply:
True.  I hope we have some limit on page sizes anyway, but I guess all of these things are terrible tools really, and all poor proxies for the real problems.  I am just suspicious of guesstimating RFs and page sizes when picking defaults.

I look forward to the day this can all be dropped in favour of proper resource management.